mm2src/coins/utxo/utxo_balance_events.rs

```diff
@@ -0,0 +1,146 @@
+
+use super::utxo_standard::UtxoStandardCoin;
+use crate::{utxo::{output_script, rpc_clients::electrum_script_hash, utxo_common::address_balance,
+                   utxo_tx_history_v2::UtxoTxHistoryOps},
+            MarketCoinOps, MmCoin};
+
+#[async_trait]
+impl EventBehaviour for UtxoStandardCoin {
+    const EVENT_NAME: &'static str = "COIN_BALANCE";
+
+    async fn handle(self, _interval: f64, tx: oneshot::Sender<EventInitStatus>) {
+        const RECEIVER_DROPPED_MSG: &str = "Receiver is dropped, which should never happen.";
+
+        let ctx = match MmArc::from_weak(&self.as_ref().ctx) {
+            Some(ctx) => ctx,
+            None => {
+                let msg = "MM context must have been initialized already.";
+                tx.send(EventInitStatus::Failed(msg.to_owned()))
+                    .expect(RECEIVER_DROPPED_MSG);
+                panic!("{}", msg);
+            },
+        };
+
+        let addresses = match self.my_addresses().await {
+            Ok(t) => t,
+            Err(e) => {
+                tx.send(EventInitStatus::Failed(e.to_string()))
+                    .expect(RECEIVER_DROPPED_MSG);
+                panic!("{}", e);
+            },
+        };
+
+        for address in addresses {
+            let script = output_script(&address, keys::Type::P2PKH);
+            let script_hash = electrum_script_hash(&script);
+            let scripthash = hex::encode(script_hash);
+
+            if let Err(e) = self
+                .as_ref()
+                .rpc_client
+                .blockchain_scripthash_subscribe(scripthash)
+                .compat()
+                .await
+            {
+                tx.send(EventInitStatus::Failed(e.to_string()))
+                    .expect(RECEIVER_DROPPED_MSG);
+                panic!("{}", e);
+            }
+        }
+
+        let scripthash_notification_receiver = match self.as_ref().scripthash_notification_receiver.as_ref() {
+            Some(t) => t,
+            None => {
+                let e = "Scripthash notification receiver can not be empty.";
+                tx.send(EventInitStatus::Failed(e.to_string()))
+                    .expect(RECEIVER_DROPPED_MSG);
+                panic!("{}", e);
+            },
+        };
+
+        tx.send(EventInitStatus::Success).expect(RECEIVER_DROPPED_MSG);
+
+        while let Some(notified_scripthash) = scripthash_notification_receiver.lock().await.next().await {
+            let address = match self.my_addresses().await {
+                Ok(addresses) => addresses.into_iter().find_map(|addr| {
+                    let script = output_script(&addr, keys::Type::P2PKH);
+                    let script_hash = electrum_script_hash(&script);
+                    let scripthash = hex::encode(script_hash);
+
+                    if notified_scripthash == scripthash {
+                        Some(addr)
+                    } else {
+                        None
+                    }
+                }),
```

Comment:
We already do scripthash calculations to trigger subscription before the loop. What do you think about creating a `scripthash->Address` map and accessing it directly? This way, we won't need to request all addresses on each notification and iterate over them.